enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,141 @@
+
+This enhancement suggests a new method for providing the self-hosted stack.
+ 
+*Note:* as per the OCP plan to move API and Ingress to K8S SLB, this direction should be investigated deeper after MetalLb will be integrated.
+
+
+## Motivation
+
+To fully control which customers run our self hosted stack and to be able to serve the needs they really have instead of providing them in a generic way. 
+Additionally, detaching the self hosted stack configuration from the selected platform will allow new customers from UPI (platform: None) to leverage the self-hosted stack.
+
+
+### Goals
+
+- Ability to use the self-hosted stack in all on-prem platforms.
+- Ability to control the self-hosted stack configuration.
+- Retrieve the self-hosted stack status through K8S API.
+- Maintain a single code base/manifests for all the on-prem platforms.
+- Use the self-hosted stack also for UPI deployments.
+
+
+### Non-Goals
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/on-prem) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by a new operator instead of static pods.
+
+The following functionalities are expected as part of the new operator:
+* Introduce a CRD used by the operator to manage the self-hosted stack.
+* Fields in CR spec that control the self-hosted stack configuration.
+* Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+* Run the self-hosted stack components (Keepalived, Coredns-MDNS, HAProxy, MDNS-publisher) as a K8S applications.
+
+### Kubelet and Operator circular dependency: 
+In order to run the operator's component, Kubelet in each node should first join the cluster by communicating with the control plane, however the components provided by the operator are responsible for enabling that.
+As a result, the below circular dependency should be addressed:
+* Kubelet can't talk to the control plane until the operator has started.
+* Operator can't start until Kubelet can talk to the control plane.
+
+To break the circular dependency, the operator should start running before bootstrap is completed in order to assure that bootstrap owns the API-VIP. In addition, a new /etc/hosts entry should be added to the nodes using MCO ignition.  This /etc/hosts entry is expected to resolve api-int hostname to the API-VIP address. 
+
+### CRD definition
+
+Below is a possible option for a CRD instance of this operator:
+
+
+```yaml
+apiVersion: clusterhosted.operator.io/v1alpha1
+kind: ClusterHostedConfig
```

Comment:
I think we'll need a more descriptive name here, e.g ClusterHosted doesn't tell us *what* is cluster hosted :)

Reply:
I'm totally open to ideas.

Reply:
ClusterHostedNetworkingConfig? openshift-hosted-network-services was one of the suggestions for the namespace where these would live in https://github.com/openshift/enhancements/pull/412.

Reply:
How about  **ClusterHostedNetServicesConfig**  ?